WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.Wallets;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class CoinJoinClient : BackgroundService, IDisposable
+	{
+		private bool _disposedValue;
+		private CredentialPool AmountCredentialPool { get; } = new();
+		private CredentialPool VsizeCredentialPool { get; } = new();
+		private Round Round { get; }
+		public IArenaRequestHandler ArenaRequestHandler { get; }
+		private BitcoinSecret BitcoinSecret { get; }
+		public KeyManager Keymanager { get; }
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+
+		public CoinJoinClient(
+			Round round,
+			IArenaRequestHandler arenaRequestHandler,
+			BitcoinSecret bitcoinSecret,
+			IEnumerable<Coin> coins,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			BitcoinSecret = bitcoinSecret;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync().ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				Money[]? othersInputAmounts = null; // TODO: Get it from somewhere.
+
+				// Planning
+				var outputs = await PlanAmountsAsync(othersInputAmounts, stoppingToken).ConfigureAwait(false);
+
+				// Output registration.
+				await RegisterOutputsAsync(outputs).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync()
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+			try
+			{
+				foreach (var coin in Coins)
+				{
+					// Parallelize or Random delay?
+					aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, BitcoinSecret, Round.Id, Round.Hash, Round.FeeRate).ConfigureAwait(false));
+				}
+			}
+			catch (Exception)
+			{
+				foreach (var alice in aliceClients)
+				{
+					// Remove already registered Inputs.
+					try
+					{
+						await alice.RemoveInputAsync().ConfigureAwait(false);
+					}
+					catch (Exception)
+					{
+						// Log?
+					}
+				}
+				throw;
+			}
```

Comment:
I think this is not what we want to do, depending the reason of the failure we should decide what to do but i believe that in most of the cases what we should do is retry. Of course if the error is a protocol error then we have a bug somewhere.

Reply:
Transport errors should be retried implicitly by the transport layer. Since we agreed to make request handling idempotent, it can do this in a fully black box manner (literally just resend the same serialized request). When the number of retries has been exceeded that shouldn't be recoverable.

I can't think of any error apart from that that shouldn't cause a failure, but the main consideration is privacy leaks through correlated deregistration - even if it makes no sense to attempt to recover from any error here i think it makes sense to proceed with the inputs that did succeed so long as the decomposition is acceptable.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/5694#discussion_r623033574